File: c++/Prologix.cpp


#include "Prologix.h"
#include <string>
#include <cstdlib>
#include <iostream>


Prologix_GPIB_USB::Prologix_GPIB_USB() {
  this->addCMD("++ver", &Prologix_GPIB_USB::version);
  this->addCMD("++addr", &Prologix_GPIB_USB::addr);
  this->addCMD("++kill", &Prologix_GPIB_USB::kill); 
}

Prologix_GPIB_USB::~Prologix_GPIB_USB() {

}

string Prologix_GPIB_USB::handleCMD(Interface *fpl,
				    string cmd,
				    vector<string> *args) {
  string out = "";
  handler h;

  if (out != "") {
    return out;
  } else if ( this->cmds.count(cmd) > 0 ) {
    h = this->cmds.at(cmd);
    return (this->*h)(fpl, args);
  }

  return "";
}

string Prologix_GPIB_USB::version(Interface *fpl,
				  vector<string> *args) {
  return "Prologix Version 1.1 Simulator (C++)";
}

string Prologix_GPIB_USB::addr(Interface *fpl,
			       vector<string> *args) {
  if (args == NULL || args->size() == 0) {
    return std::to_string(fpl->addr);
  } else {
    fpl->addr = atoi(args->at(0).c_str());
    if ( fpl->devices.count(fpl->addr) == 0 )
      fpl->addDevice(this, fpl->addr);
    return "";
  }
}

string Prologix_GPIB_USB::kill(Interface *fpl,
			       vector<string> *args) {
  fpl->running = false;
  return "";
}


void Prologix_GPIB_USB::addCMD(const char *cmd, handler h) {
  this->cmds.insert( std::pair<string, handler>(cmd, h) );
}

void Prologix_GPIB_USB::addCMD(string cmd, handler h) {
  this->cmds.insert( std::pair<string, handler>(cmd, h) );
}

File: python/Prologix.py
##
## **** Prologix_GPIB_USB Class ****
##
## A Device that emulates the functionality of the
## Prologix GPIB-USB Controller (prologix.biz)
## 


import Device, PTY_Interface

class Prologix_GPIB_USB(Device.Device):

##%% Initialization - Prologix_GPIB_USB()
## Does not take arguments, but is VERY important
## All classes that attempt to inherit this class,
## must call this initialization.
##
## This creates a dictionary containing all the commands
## to be run. If this dictionary is not created, this class
## does nothing
##
    def __init__(self):
        self.cmds = { "++ver":self.version, # "COMMAND":HANDLER
                      "++addr":self.addr,
                      "++kill":self.kill}

##%% handleCMD(pty_interface, command, arguments)
## See Device.handleCMD documentation
##
## In Prologix_GPIB_USB, this function makes use of the
## map made in initialization to find the command and
## its associated handler, then call the handler and pass in
## the pty_interface and the arguments
##
    def handleCMD(self, fpl, cmd, args=[]):
        out = Device.Device.handleCMD(self, fpl, cmd, args)
        if out != "":
            return out
        elif cmd in self.cmds:
            return self.cmds[cmd](fpl, args)
        else:
            return ""

##%% version(pty, args)
## Returns the device version
##
    def version(self, fpl, args=[]):
        return "Prologix Version 1.1 Simulator (Python)"

##%% addr(pty, args)
## If no arguments are given, returns the current address
## of the pty interface
##
## if args are given, then an integer is parsed out of
## the first argument and set as the address. This ensures
## that a device exists at the following address
##
## Any errors return "ERROR"
##
    def addr(self, fpl, args=[]):
        if args == []:
            return str(fpl.addr)
        else:
            try:
                fpl.addr = int(args[0])
                if not fpl.addr in fpl.devices:
                    fpl.addDevice(PrologixAdapter(), fpl.addr)
                return ""
            except:
                return "ERROR"
##%% kill(pty, args)
## Stops the pty interface from running
##     
    def kill(self, fpl, args=[]):
        fpl.running = False
        return ""

##%% addCMD(command, handler_function)
## Adds a command to the dictionary
## Handler function must be of the form
## func(pty_interface, arguments):
##
    def addCMD(self, cmd, func):
        self.cmds[cmd] = func

